### PR TITLE
Airport-Challenge-Jenna

### PR DESCRIPTION
**Pull request by Jen-Chap**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,70 +12,193 @@ Airport Challenge
...
+|         |airportName@String             |                   |         |
+| Plane   |planeID @String                |                   |         |
+
+**Tests**  
+*Test 1* - Check that only Plane types are able to land at the Airport when planeLands  
+*Test 2* - Length of planesAtAirport array at the Airport increases by 1 when planeLands  
+*Test 3* - Check that edge cases such as 'null' as a type are unable to land at the airport and length of planesAtAirport remains at 0  
+*Test 4* - Check that when something that isn't a plane attempts to land at the airport, a message is returned  
+*Test 5* - When a new airport instance is created from the Airport class, it contains an airportName property  
+*Test 6* - When a new plane instance is created from the plane class, it contains a planeID property
```
Comment (Jen-Chap) [inline README.md:34]: Should I have a created another test that checks planeID property is undefined when an ID is not provided? Is there a  way to block the creation of an instance if a value is not given?

File: README.md [hunk 2]
```diff
@@ -12,70 +12,193 @@ Airport Challenge
...
+
+**Tests**  
+
+*Test 1* - Check the value returned for default airport capacity  
+*Test 2* - Check the value returned when capacity changed with positive integer  
+*Test 3* - Return an error message when capacity is inputted as below 0  
+*Test 4* - Return an error message when capacity is inputted as null  
+*Test 5* - Return an error message when capacity is inputted as a non-numerical string  
+  
+To note: For tests 2.3 - 2.5, I wanted to throw an error rather than set the capacity as a string, however I then couldn't run the tests.
```
Comment (Jen-Chap) [inline README.md:58]: Is the throw new error only used when testing rather than in the actual code?

File: README.md [hunk 3]
```diff
@@ -12,70 +12,193 @@ Airport Challenge
...
+|         |airportName@String            |                   |         |
+|         |airportCapacity @Integer      | atCapacity()      |@Boolean |
+| Plane   |plane id @String              |                   |         |
+
+**Tests**  
+
+*Test 1* - When the length of planesAtAirport and airportCapacity are the same, the test returns true  
+*Test 2* - When the length of planesAtAirport and airportCapacity are the same (1), atCapacity returns true  
+*Test 3* - When the length of planesAtAirport and airportCapacity are different, atCapacity returns false  
+*Test 4* - When atCapacity is true, planeLands returns an error message
```
Comment (Jen-Chap) [inline README.md:79]: I should have added a test for the atCapacity function if an incorrect value had been inputted for airportcapacity and so was set as a string error message

File: README.md [hunk 4]
```diff
@@ -12,70 +12,193 @@ Airport Challenge
...
-
-As an air traffic controller
-To count planes easily
-Planes that have landed must be at an airport
-```
-
-Your task is to test drive the creation of a set of classes/objects to satisfy all the above user stories. You will need to use a random number generator to set the weather (it is normally sunny but on rare occasions it may be stormy). In your tests, you'll need to stub random behaviour to ensure consistent test behaviour.
-
-Your code should defend against [edge cases](http://programmers.stackexchange.com/questions/125587/what-are-the-difference-between-an-edge-case-a-corner-case-a-base-case-and-a-b) such as inconsistent states of the system ensuring that planes can only take off from airports they are in; planes that are already flying cannot take off and/or be in an airport; planes that are landed cannot land again and must be in an airport, etc.
+## User Story 1
```
Comment (Jen-Chap) [inline README.md:15]: General comments:
- I would encapsulate the properties within Airport and Plane and create getters and setters now
- I would test with Jasmine so I could use SpyOn to check if functions had been called

File: README.md [hunk 5]
```diff
@@ -12,70 +12,193 @@ Airport Challenge
...
+|         |                              | planeTakesOff()                   | @String |
+|         |                              | isPlaneAtAirport(@Plane)          | @Boolean|
+| Plane   |planeID @String               |                                   |         |
+
+**Tests**  
+
+*Test 1* - Length of planesAtAirport array at the Airport decreases by 1 when planeTakesOff  
+*Test 2* - Confirm if plane is in planesAtAirport using isPlaneAtAirport returns true  
+*Test 3* - Confirm if plane is in planesAtAirport using isPlaneAtAirport returns false  
+*Test 4* - When planeTakesOff is called and isPlaneAtAirport returns true, length of planesAtAirport decreases by 1
```
Comment (Jen-Chap) [inline README.md:104]: Do I need this test (and similar ones below) where I check the length of planesAtAirport again as I've created a new method/changed something? Or would the original passing (4.1) be enough?

File: README.md [hunk 6]
```diff
@@ -12,70 +12,193 @@ Airport Challenge
...
+| Airport |planesAtAirport @Array[@Plane] | planeLands(@Plane)                | @Void   |
+|         |airportName@String             |                                   |         |
+|         |airportCapacity @Integer       | atCapacity()                      | @Boolean|
+|         |                               | planeTakesOff(@Plane)             | @Void   |
+|         |                               | planeIsAtAirport(@Plane)          | @Boolean|
+| Plane   |planeID  @String               |                                   |         |
+
+**Tests**  
+
+*Test 1* - When planeTakesOff is called and isPlaneAtAirport returns False, an error message is returned
```
Comment (Jen-Chap) [inline README.md:124]: I'm not sure if this is the best way to return an error message and as in the future you may want the message to change, I could have tested by seeing is a string type was returned rather than the specific message

File: README.md [hunk 7]
```diff
@@ -12,70 +12,193 @@ Airport Challenge
...
+|         |                               | planeTakesOff(@Plane, @Weather)   | @Void   |
+|         |                               | planeIsAtAirport(@Plane)          | @Boolean|
+| Plane   |planeID  @String               | planeLocation(@airport)           | @String |
+|         |location @string               |                                   |         |
+| Weather |currentWeather @String         | checkWeather()                    |         |
+|         |possibleWeather @Array         |                                   |         |
+
+**Tests**
+
+Note: I think this User Story has already been addressed and tested with the 'planeIsAtAirport' function, however I've added a function to the plane class to return if it's at a specific airport (this could be expanded to a list of airports with a for loop) or in flight
```
Comment (Jen-Chap) [inline README.md:201]: To expand, the options I considered were:
- Creating an index.js file which contained the method to call planeIsAtAirport and check the length of planesAtAirport array
- Adding a line within planeLands that calls planeIsAtAirport to return true
- The method I chose to add (planeLocation) I think I should be called within planeLands so the planes location changes, however this could get messy and is probably not the best way to achieve this (or the user story)

File: README.md [hunk 8]
```diff
@@ -12,70 +12,193 @@ Airport Challenge
...
+|         |location @string               |                                   |         |
+| Weather |currentWeather @String         | checkWeather()                    |         |
+|         |possibleWeather @Array         |                                   |         |
+
+**Tests**
+
+Note: I think this User Story has already been addressed and tested with the 'planeIsAtAirport' function, however I've added a function to the plane class to return if it's at a specific airport (this could be expanded to a list of airports with a for loop) or in flight  
+  
+*Test 1* - Plane location returns 'In flight' by default  
+*Test 2* - planeLocation changes location if plane is in planesAtAirport when an airport is inputted
```
Comment (Jen-Chap) [inline README.md:204]: I wrote these tests into the Airport spec but they should be in the plane spec

File: src/Plane.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+class Plane {
+    constructor(planeID, location = 'In flight') {
+        this.planeID = planeID;
+        this.location = location
+    }
+
+    planeLocation = (airport) => {
+        if (airport.planesAtAirport.includes(this.planeID)) {
+            return this.location = airport.name;
+        }
```
Comment (Jen-Chap) [inline src/Plane.js:10]: If I were to expand on this method, I would need an else statement to change location back to 'In flight' if it wasn't in a list of airport names (I would have to input a list of airports as an array and use for loops)

File: src/Weather.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class Weather {
+    currentWeather;
+    possibleWeather = ['sunny', 'sunny', 'stormy',]
+
+    checkWeather = () => {
+        let index = Math.floor(Math.random() * this.possibleWeather.length);
+        return this.currentWeather = this.possibleWeather[index]
```
Comment (Jen-Chap) [inline src/Weather.js:7]: Which is better, the two lines above or:
return this.currentWeather = this.possibleWeather[Math.floor(Math.random() * this.possibleWeather.length)]

the lines written were easier for me to read and understand but would it be better to combine them into one line?

File: src/airport.js [hunk 11]
```diff
@@ -1,5 +1,34 @@
+const Plane = require('./Plane.js') //only used to check instance of - is there a way around this?
+
 class Airport {
-  // here's a starting point for you
+  planesAtAirport = []
+
+  constructor(name, airportCapacity = 10) {
+    if (typeof (airportCapacity === 'number') && airportCapacity > 0) {
+      this.airportCapacity = airportCapacity
+    } else { this.airportCapacity = "Capacity not set, must be a positive integer" };
```
Comment (Jen-Chap) [inline src/airport.js:9]: Is there a way to prevent an instance being made in this case?

File: src/airport.js [hunk 12]
```diff
@@ -1,5 +1,34 @@
...
+    this.name = name;
+  };
+
+  atCapacity = () => {
+    return this.planesAtAirport.length === this.airportCapacity ? true : false
```
Comment (Jen-Chap) [inline src/airport.js:14]: I would check if I need the '? true: false' or would it automatically return true and false anyway?

File: tests/Airport.spec.js [hunk 13]
```diff
@@ -0,0 +1,792 @@
+const { assertEquals } = require('./testingFrameworks.js');
+const Airport = require('../src/Airport');
+const Plane = require('../src/Plane')
+const Weather = require('../src/Weather')
+
+let expectedOutput;
+let actualOutput, result;
+let airport1 = new Airport()
+let plane1 = new Plane()
```
Comment (Jen-Chap) [inline tests/Airport.spec.js:9]: I would name these testAirport and testPlane

File: tests/Airport.spec.js [hunk 14]
```diff
@@ -0,0 +1,792 @@
...
+
+// Report
+console.log(`Plane landed: ${result}`);
+
+// Clean-Up
+actualOutput = undefined;
+result = undefined;
+expectedOutput = undefined;
+
+// Test 6.1
```
Comment (Jen-Chap) [inline tests/Airport.spec.js:551]: 6.1 and 6.2 should be in a separate weather spec
